src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php

```diff
@@ -0,0 +1,505 @@
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Class CurrencyController is responsible for handling "Improve -> International -> Localization -> Currencies" page.
+ */
+class CurrencyController extends FrameworkBundleAdminController
+{
+    /**
+     * Show currency page.
+     *
+     * @AdminSecurity(
+     *     "is_granted('read', request.get('_legacy_controller'))",
+     *     redirectRoute="admin_currencies_index"
+     * )
+     *
+     * @param CurrencyFilters $filters
+     *
+     * @return Response
+     */
+    public function indexAction(CurrencyFilters $filters)
+    {
+        $currencyGridFactory = $this->get('prestashop.core.grid.factory.currency');
+        $currencyGrid = $currencyGridFactory->getGrid($filters);
+        $gridPresenter = $this->get('prestashop.core.grid.presenter.grid_presenter');
+
+        $settingsForm = $this->getSettingsFormHandler()->getForm();
+
+        return $this->render('@PrestaShop/Admin/Improve/International/Currency/index.html.twig', [
+            'currencyGrid' => $gridPresenter->present($currencyGrid),
+            'currencySettingsForm' => $settingsForm->createView(),
+        ]);
+    }
+
+    /**
+     * Provides filters functionality.
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function searchAction(Request $request)
+    {
+        $definitionFactory = $this->get('prestashop.core.grid.definition.factory.currency');
+        $definitionFactory = $definitionFactory->getDefinition();
+
+        $gridFilterFormFactory = $this->get('prestashop.core.grid.filter.form_factory');
+        $searchParametersForm = $gridFilterFormFactory->create($definitionFactory);
+        $searchParametersForm->handleRequest($request);
+
+        $filters = [];
+        if ($searchParametersForm->isSubmitted()) {
+            $filters = $searchParametersForm->getData();
+        }
+
+        return $this->redirectToRoute('admin_currencies_index', ['filters' => $filters]);
+    }
+
+    /**
+     * Displays and handles currency form.
+     *
+     * @AdminSecurity(
+     *     "is_granted('create', request.get('_legacy_controller'))",
+     *     redirectRoute="admin_currencies_index",
+     *     message="You do not have permission to add this."
+     * )
+     *
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function createAction(Request $request)
+    {
+        $multiStoreFeature = $this->get('prestashop.adapter.multistore_feature');
+
+        $currencyForm = $this->getCurrencyFormBuilder()->getForm();
+        $currencyForm->handleRequest($request);
+
+        try {
+            if ($currencyForm->isSubmitted() && $this->isDemoModeEnabled()) {
+                $this->addFlash('error', $this->getDemoModeErrorMessage());
```

Comment:
if i remember correctly, demo mode is handled in form handling system, so you should not worry too much about it.